contracts/Utils.sol

```diff
@@ -23,26 +38,29 @@ contract Utils {
 
     mapping(IERC20 => uint256) internal decimals;
 
-    function getUpdateDecimals(IERC20 token) internal returns (uint256) {
-        if (token == ETH_TOKEN_ADDRESS) return ETH_DECIMALS; // save storage access
-        uint256 tokenDecimals = decimals[token];
-        // moreover, very possible that old tokens have decimals 0
-        // these tokens will just have higher gas fees.
-        if (tokenDecimals == 0) {
-            tokenDecimals = token.decimals();
-            decimals[token] = tokenDecimals;
-        }
-
-        return tokenDecimals;
+    function getSetDecimals(IERC20 token) internal returns (uint256 tokenDecimals) {
+        tokenDecimals = getDecimals(token);
+        if (getDecimalsConstant(token) == 0) decimals[token] = tokenDecimals;
```

Comment:
we should set decimals[token] only if wasn't  already set.
since we pay extra gas here when not required.